### ISSUE TITLE
side effect in useMemoOne

### ISSUE THREAD

**Issue by auver** [alexreardon/use-memo-one#21]
https://codesandbox.io/s/condescending-poitras-zp2t7?file=/src/App.js

```js
import React, { useState, useMemo } from "react";
import { useMemoOne } from "use-memo-one";
import "./styles.css";

export default function App() {
    const [state, setState] = useState(0);

    const [count, setCount] = useState(0);

    useMemoOne(() => {
        setState(count + 10);
    }, [count]);
    return (
        <div className="App">
            <h1>{count}</h1>
            <h2>{state}</h2>
            <button onClick={() => setCount((c) => c + 1)}>setCount</button>
        </div>
    );
}
```

If it has some side effect in useMemoOne, like set other state, it makes App re-render. And this useMemoOne will run again though count if not changed, which causes the infinite loop.

Use `React.useMemo` will be OK.

**Comment by alexreardon** [comment]
I think this behaviour is expected

**Comment by alexreardon** [comment]
the `useMemoOne` function is trying to update the cached value based on the `inputs` array. In this case the memoized function is causing a side effect that is breaking the 'cache' of the memoized function, which in turn is making the infinite loop. I wonder what the behaviour of `useMemo` is in this case

**Comment by gmoniava** [comment]
@alexreardon why there is no infinite loop with `useMemo`, any ideas?

**Comment by auver** [comment]
One possible way is commit the cache synchronously  instead of `useEffect`

**Comment by gmoniava** [comment]
>  causing a side effect that is breaking the 'cache' of the memoized function


@alexreardon But `setState` doesn't modify `count` which is dependency for `useMemo`, so why is there infinite loop after click? How does it break the cache as you mention?

**Comment by alexreardon** [comment]
I'll try to take a look at this soon

**Comment by gmoniava** [comment]
@alexreardon I looked into your code the other day but didn't notice why there is infinite loop after click, just by looking at it. Didn't have time yet, to debug through it. Feel free to share reason why that happens if you find out.

**Comment by gmoniava** [comment]
@alexreardon Problem happens here:

```
const cache = useCache
    ? committed.current
    : {
        inputs,
        result: getResult()
      };
```

When user clicks the button, `useCache `is false the first time, and it runs `getResult()` from above, that, in turn cause a re-render,
which calls the custom hook again - the thing is in this process the `useEffect `never fires. Hence, the infinite loop happens.

**Comment by miuirussia** [comment]
According to react documentation function inside useMemo must be pure (without side effects), so render loop expected and it does not need to be fixed

**Comment by gmoniava** [comment]
@miuirussia Could be true, but I could not find where react docs say that.

**Comment by miuirussia** [comment]
https://reactjs.org/docs/hooks-reference.html#usememo

> Remember that the function passed to useMemo runs during rendering. Don’t do anything there that you wouldn’t normally do while rendering. For example, side effects belong in useEffect, not useMemo.

**Comment by gmoniava** [comment]
@miuirussia Hmm it's not entirely clear what kind of purity they refer to, for example I can do a `setState` **inside** render given I do it in condition, like:
```

if(something){
 setState(1);
}
```

**Comment by miuirussia** [comment]
But these are bad practices

**Comment by gmoniava** [comment]
@miuirussia The example from my previous comment was based on react docs, check the link.

**Comment by miuirussia** [comment]
Yes, but it is clearly written there that you should not do this

**Comment by gmoniava** [comment]
@miuirussia No it doesn't say that. If you refer to this:

> While you probably don’t need it, in rare cases that you do (such as implementing a <Transition> component), you can update the state right during rendering

it is talking about `getDerivedStateFromProps` not about calling setState during render.

**Comment by gmoniava** [comment]
@miuirussia you could add comment what you disagree with instead of downvote ;)
